proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+  with the additional info needed for compilation, e.g. used external libraries and properties that should be visible 
+  from the script; in addition it may extract relevant part of the script for partial or staged script handling (e.g. as
+  the dependencies/plugins sections in gradle).
+  
+#### Various script types
+
+There could be a need to support various types of the scripts in the same environment. E.g. in a web framework there 
+could be routing scripts and content scripts. For such cases one more component should be configured:
+- *Selector* - the service that recognizes a script using file name or text
+
+The same component could be used for script type selection in an IDE.
+
+#### Caching
+
+Since calling Kotlin compiler could be quite a heavy operation in comparison with typical script execution, the caching
+of the compiled script should be supported by the compilation platform.
+
+#### Execution lifecycle
+
+The script is executed in the following pipeline:
+- selection - the *Selectors* are called on the script to choose the set of services that will handle it
+- preprocessing - the *Preprocessor* takes the script and the environment and extracts the compilation configuration
+- compilation - the *Platform* takes the script, and the compilation configuration and provides the compiled class
+- instantiation - the *Executor* takes the compiled class, the environment, and the compilation configuration and
+  instantiates the script object, passing the required parameters from the environment to the constructor
+- execution - the *Executor* takes the instantiated script and executes the appropriate method, passing arguments
+  from the environment to it; this step could be repeated many times
+
+#### Processing in an IDE
+
+The script is handled in an IDE in the following sequence (*Note that the first two steps are the same as in the 
+execution pipeline*):
+- selection - the *Selectors* takes the script and selects the set of services that will handle it
+- preprocessing - the *Preprocessor* takes the script and the environment and extracts the compilation configuration
+- resolution info extraction - the *Platform* takes the script and the compilation configuration and provides the
+  info required for rich editing the script
+
+### Standalone scripting
+
+Standalone scripting applications include command-line utilities and a standalone Kotlin REPL.
+
+Standalone scripting is a variant of the embedded scripting with hosts provided with the Kotlin distribution.
+
+#### Hosts
+
+The standalone script could be executed e.g. using command line Kotlin compiler:
+
+`kotlinc -script myscript.kts`
+
+Or with some hypothetical dedicated runner, that we might want to include into the distribution:
+
+`kotlinsh myscrtipt.ktsh`
+
+To be able to use the Kotlin scripts in a Unix shell environment, the *shebang* (`#!`) syntax should be supported 
+at the beginning of the script:
+```
+#! /path/to/kotlin/script/runner -some -params
```

Comment:
Yes, thanks, I remember that. Therefore the syntax of passing params in the arguments to the annotations should be supported as well.